tests/simulator_test.cpp

```diff
@@ -2,82 +2,269 @@
 #include "main/simulator.h"
 #include "gtest/gtest.h"
 
-// ! OLD ENERGY COLLECTORS TEST
-/*
-TEST_F(SceneManagerTest, EnergyCollectorPositionsCheck) {
-  for (const auto &collectorPtr : manager.getEnergyCollectors()) {
-    std::cout << *collectorPtr << std::endl;
+using constants::kPi;
+using core::Ray;
+using core::RayHitData;
+using core::Vec3;
+
+const int kSkipNumCollectors = 37;
+const float kSkipFrequency = 1000;
+const Vec3 kVecZero(0, 0, 0);
+const Vec3 kVecUp(0, 0, 1);
+const Vec3 kVecX(1, 0, 0);
+const Vec3 kVecY(0, 1, 0);
+
+float deg2rad(float deg) { return 2 * kPi * deg / 360; }
+
+// TODO: add collision point check for every ray hit
+
+class FakeModel : public ModelInterface {
+
+public:
+  explicit FakeModel(bool empty) : empty_(empty){};
+  std::vector<objects::TriangleObj *> triangles() const override { return {}; }
+  float height() const override { return 0; }
+  float sideSize() const override { return 0; }
+  bool empty() const override { return empty_; }
+
+private:
+  bool empty_;
+};
+
+class EnergyCollectorTest : public ::testing::Test {
+protected:
+  // performs ray hit at at given energy collectors, |hitData| is modified to
+  // hold information where ray hit energyCollector. Returns true if hit
+  // occurred, false when there was no hit
+  [[nodiscard]] bool performHitCollector(
+      const std::vector<std::unique_ptr<objects::EnergyCollector>>
+          &energyCollectors,
+      const Ray &ray, RayHitData *hitData) {
+
+    for (const auto &collector : energyCollectors) {
+      if (collector->hitObject(ray, kSkipFrequency, hitData)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  void
+  printCollectors(const std::vector<std::unique_ptr<objects::EnergyCollector>>
+                      &energyCollectors) const {
+    for (const auto &collector : energyCollectors) {
+      std::cout << "Origin: " << collector->getOrigin() << "\n"
+                << "Radius: " << collector->getRadius() << "\n";
+    }
+  }
+};
+TEST_F(EnergyCollectorTest, ThrowExceptionWhenInvalidNumCollector) {
+
+  const FakeModel nonEmptyModel(false);
+
+  const int invalidNumCollectors = 38;
+  try {
```

Comment:
Seriously? :-)  In the links I sent you they used macros so you can use a macro as well? Please please please, for nth time, please reuse the code you write.

Reply:
done :D